src/lua/compat.lua

```diff
@@ -0,0 +1,307 @@
+            action = ACT_DEFAULT
+        elseif mode == 'current' then
+            action = ACT_CURRENT
+        elseif mode == nil then
+            action = ACT_NIL
+        else
+            error("usage: compat.dump('new'/'old'/'default'/nil)")
+        end
+        local result = [[require('tarantool').compat({]]
+        local max_key_len = 0
+        for _, key in pairs(options_order) do
+            max_key_len = (#key > max_key_len) and #key or max_key_len
+        end
+        for _, key in pairs(options_order) do
+            local val
+            local comment
+            if options[key].obsolete then
+                if action == ACT_NEW or action == ACT_OLD or action == ACT_CURRENT or
+                        (action == ACT_NIL and options[key].selected) then
+                    val = "'new'"
+                else
+                    val = "'default'"
+                end
+                comment = 'obsolete since ' .. options[key].obsolete
+            else
+                if action == ACT_CURRENT then
+                    if options[key].selected and options[key].current == NEW or options[key].default == 'new' then
+                        val = "'new'"
+                    else
+                        val = "'old'"
+                    end
+                elseif action == ACT_DEFAULT or (action == ACT_NIL and not options[key].selected) then
+                    val = "'default'"
+                elseif action == ACT_NEW or (action == ACT_NIL and options[key].current == NEW) then
+                    val = "'new'"
+                else
+                    val = "'old'"
+                end
+            end
+            -- Can't use '\t' due to gh-7681.
+            result = ('%s\n    %s%s = %s,'):format(result, key, string.rep(' ', max_key_len - #key), val)
+            if comment then
+                if (val ~= "'default'") then
+                    -- For alignment.
+                    result = result .. '    '
+                end
+                result = result .. ' -- ' .. comment
+            end
+        end
+        if not result:find('\n') then
+            return result .. '})'
+        else
+            -- Need double '\n' due to gh-3012.
+            return result .. '\n})\n\n'
+        end
+    end,
+    add_option = function(policy)
+        if type(policy) ~= 'table' then
+            error("usage: compat.add_option({name = '...', default = 'new'/'old'," ..
+                  " brief = '...', action = func, run_action = true/false})")
+        end
+        local name = policy.name
+        verify_option(name, policy)
+
+        local current = policy.default == 'new'
+        if not options[name] then
+            options[name] = {
+                current = current,
+                selected = false
+            }
+            table.insert(options_order, name)
+        elseif not options[name].selected then
+            options[name].current = current
+        end
```

Comment:
Don't know, it is a bit complex for me. The same block incapsulates a logic about an exist/a non exist option and a selected/a non selected option.

I tried to express it in some other way. Don't know, whether the snippets below are actually better, but I would at least appreciate a comment for this block.

```lua
-- Assign a number to a newly added option.
-- It is used for ordering in dump/serialize.
local is_exists = option[name] ~= nil
if not is_exists then
    table.insert(options_order, name)
end

-- Keep current/selected values if the option
-- is already exists. Assign a default otherwise.
local option = options[name] or {
    current = policy.default == 'new',
    selected = false,
}
options[name] = option

-- If the default was changed (say, a newer version
-- of the module is loaded) and the option was not
-- explicitly set, then update option's current value
-- to the new default.
if not option.selected then
    option.current = policy.default == 'new'
end

option.brief = policy.brief
<...>
```

```lua
local is_exists = option[name] ~= nil
if not is_exists then
    -- Assign a number to a newly added option.
    -- It is used for ordering in dump/serialize.
    table.insert(options_order, name)

    -- Assign a default value.
    options[name] = {
        current = policy.default == 'new',
        selected = false,
    }
end

local option = options[name]

-- If the default was changed (say, a newer version
-- of the module is loaded) and the option was not
-- explicitly set, then update option's current value
-- to the new default.
if not option.selected then
    option.current = policy.default == 'new'
end

option.brief = policy.brief
<...>
```

Reply:
Waiting for more decent @igormunkin glance.

Reply:
Added more comments.

Reply:
Okay for me now.